### PR TITLE
Add MongoDB extension for JUnit 5

### LINKED ISSUE #7655: Allow JUnit 5 integration DB integration tests
We currently only have implemented JUnit 4 integration for MongoDB and Elastisearch integration tests. 

JUnit integration and container lifecycle management are implemented in the same class for MongoDb and Elasticsearch respectively: 
- https://github.com/Graylog2/graylog2-server/blob/86bef8d8e66229019d99d16a65b56d287cad0be9/graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBInstance.java
- https://github.com/Graylog2/graylog2-server/blob/037ff14dfc97fef886aafd1961c84e459305f976/graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchInstance.java

## Possible Solution
Lifecycle management and JUnit integration could be implemented in different classes to allow for composing them together. So we would end up with 3 small classes per DB type instead of one big one, e.g.:
- MongoDBJunit4Rule
- MongoDBJUnit5Extension
- MongoDBInstance

### PR DESCRIPTION
**Pull request by bernd**
Introduce MongoDBTestService and MongoDBContainer to remove MongoDB
service specific code from the JUnit 4 MongoDBInstance rule.

Add a JUnit 5 extension for MongoDB that is using the new
MongoDBTestService.

The MongoDBInstance API has been kept to avoid changing tons of tests.
It can go once we moved all test over to JUnit 5.

Closes #7655

### CODE REVIEW

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+            final MongoDBFixtures fixtureFiles = fixtureAnnotation.get();
+            LOG.debug("Loading fixtures {} for {}#{}()",
+                    fixtureFiles.value(),
+                    invocationContext.getTargetClass().getCanonicalName(),
+                    invocationContext.getExecutable().getName());
+            final MongoDBFixtureImporter fixtureImporter = new MongoDBFixtureImporter(fixtureFiles.value(), invocationContext.getTargetClass());
+
+            fixtureImporter.importResources(getInstance(context).mongoConnection().getMongoDatabase());
+        }
+    }
```
Comment (alex-konn) [inline graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java:189]: How about extracting private methods for finding the annotation and loading the fixtures? Would make this block a tad more readable IMO. 

```suggestion
    private void processFixtures(ReflectiveInvocationContext<Method> invocationContext, ExtensionContext context) {
        findFixtureAnnotation(context)
                .ifPresent(annotation -> loadFixtures(invocationContext, context, annotation));
    }

    private Optional<MongoDBFixtures> findFixtureAnnotation(ExtensionContext context) {
        ExtensionContext currentContext = context;
        Optional<MongoDBFixtures> fixtureAnnotation;

        // Find fixture annotation on current element or parent elements
        do {
            fixtureAnnotation = findAnnotation(currentContext.getElement(), MongoDBFixtures.class);

            if (!currentContext.getParent().isPresent()) {
                break;
            }

            currentContext = currentContext.getParent().get();
        } while (!fixtureAnnotation.isPresent() && currentContext != context.getRoot());
        return fixtureAnnotation;
    }

    private void loadFixtures(ReflectiveInvocationContext<Method> invocationContext, ExtensionContext context, MongoDBFixtures fixtureAnnotation) {
        LOG.debug("Loading fixtures {} for {}#{}()",
                fixtureAnnotation.value(),
                invocationContext.getTargetClass().getCanonicalName(),
                invocationContext.getExecutable().getName());
        final MongoDBFixtureImporter fixtureImporter = new MongoDBFixtureImporter(fixtureAnnotation.value(), invocationContext.getTargetClass());

        fixtureImporter.importResources(getInstance(context).mongoConnection().getMongoDatabase());
    }
```

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionTest.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+class MongoDBExtensionTest {
+    @Nested
+    class ProgrammaticInstanceRegistration {
+        @SuppressWarnings("unused")
+        @RegisterExtension
+        MongoDBExtension mongodbExtension = MongoDBExtension.create();
+
+        @Test
+        @MongoDBFixtures("MongoDBExtensionTest-1.json")
+        void test(MongoDBTestService mongodb) {
```
Comment (alex-konn) [inline graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionTest.java:40]: I think it makes sense to give all tests meaningful names. I would call this one `importsCollectionFromFixture` or something similar.

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionWithRegistrationAsStaticFieldTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+import static com.mongodb.client.model.Filters.eq;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Testing programmatic extension registration with static fields has been moved to this separate file because
+ * we cannot use static fields on nested classes in {@link MongoDBExtensionTest}.
+ */
+// Required to make the instance ID check work
+@TestMethodOrder(OrderAnnotation.class)
```
Comment (alex-konn) [inline graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionWithRegistrationAsStaticFieldTest.java:39]: I think the test could be improved by moving the tests asserting that the same instance id is used to a nested class. Running tests in order is only relevant for them so `@TestMethodOrder` would only be needed on the nested class and we could also get rid of the `Order` annotation on the last two tests.
The static field for `instanceIds` would still have to go in the outer class, which is a bit ugly, but a trade off, I'd be fine with.

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionWithRegistrationAsStaticFieldTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    void registerInstanceId1(MongoDBTestService mongodb) {
+        // Just add the MongoDB instance ID so we can check them in checkInstanceIds()
+        instanceIds.add(mongodb.instanceId());
+    }
+
+    @Test
+    @Order(2)
+    void registerInstanceId2(MongoDBTestService mongodb) {
+        // Just add the MongoDB instance ID so we can check them in checkInstanceIds()
+        instanceIds.add(mongodb.instanceId());
```
Comment (alex-konn) [inline graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionWithRegistrationAsStaticFieldTest.java:70]: Wouldn't it be enough to already assert here that the injected instance's id is equal to the one injected in the test before?

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBInstance.java [hunk 5]
```diff
@@ -49,27 +41,19 @@
  */
 public class MongoDBInstance extends ExternalResource implements AutoCloseable {
```
Comment (alex-konn) [inline graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBInstance.java:42]: From my point of view, we can also get rid of the unused `createForEachTest` method.

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBTestService.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+        this.mongoConnection = new MongoConnectionImpl(mongoConfiguration);
+        this.mongoConnection.connect();
+        this.mongoConnection.getMongoDatabase().drop();
+    }
+
+    /**
+     * Close (shutdown) the service.
+     *
+     * @throws Exception when closing the service fails
```
Comment (alex-konn) [inline graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBTestService.java:110]: Does it? ;)

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    /**
+     * Create new extension instance using the given MongoDB service version.
+     *
+     * @return the new extension instance
+     */
+    public static MongoDBExtension create(String version) {
+        return new MongoDBExtension(requireNonNull(version, "version cannot be null"));
+    }
+
+    public MongoDBExtension() {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java:80]: This is used by the JUnit 5 extension framework when using `@ExtendWith`. I suppressed the warning and added a comment.

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionTest.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        @MongoDBFixtures("MongoDBExtensionTest-1.json")
+        void withMethodAnnotation(MongoDBTestService mongodb) {
+            // Is ONLY using the method annotation and doesn't load the fixture from the class annotation
+            assertEmptyCollection(mongodb, "test_3");
+            assertCollection(mongodb, "test_2", eq("field_a", "content1"), "54e3deadbeefdeadbeef0001");
+        }
+    }
+
+    private MongoCollection<Document> mongoCollection(MongoDBTestService mongodb, String name) {
+        return mongodb.mongoConnection().getMongoDatabase().getCollection(name);
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionTest.java:101]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoDBExtension.class);
+    private static final ExtensionContext.Namespace NAMESPACE = ExtensionContext.Namespace.create(MongoDBExtension.class);
+
+    /**
+     * Create new extension instance using default settings.
+     *
+     * @return the new extension instance
+     */
+    public static MongoDBExtension create() {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java:67]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+            return MongoDBTestService.create(version);
+        });
+    }
+
+    private void closeInstance(ExtensionContext context) throws Exception {
+        context.getStore(NAMESPACE).remove(Lifecycle.class);
+        getInstance(context).close();
+    }
+
+    private void clearInstance(ExtensionContext context) throws Exception {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java:103]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBTestService.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+     */
+    @Override
+    public void close() throws Exception {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBTestService.java:113]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+        if (context.getStore(NAMESPACE).get(Lifecycle.class) == null) {
+            context.getStore(NAMESPACE).put(Lifecycle.class, lifecycle);
+        }
+        return (MongoDBTestService) context.getStore(NAMESPACE).getOrComputeIfAbsent(MongoDBTestService.class, c -> {
+            LOG.debug("Starting a new MongoDB service instance with lifecycle {}", lifecycle);
+            return MongoDBTestService.create(version);
+        });
+    }
+
+    private void closeInstance(ExtensionContext context) throws Exception {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java:98]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBTestService.java [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+public class MongoDBTestService implements AutoCloseable {
+    private static final Logger LOG = LoggerFactory.getLogger(MongoDBTestService.class);
+
+    private static final String DEFAULT_DATABASE_NAME = "graylog";
+    public static final String DEFAULT_VERSION = MongoDBContainer.DEFAULT_VERSION;
+
+    private final MongoDBContainer container;
+    private MongoConnectionImpl mongoConnection;
+
+    public static String defaultVersion() {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBTestService.java:42]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionTest.java [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+        @Test
+        @MongoDBFixtures("MongoDBExtensionTest-1.json")
+        void test(MongoDBTestService mongodb) {
+            assertCollection(mongodb, "test_1", eq("hello", "world"), "54e3deadbeefdeadbeef0000");
+        }
+
+        @Test
+        void withoutFixtures(MongoDBTestService mongodb) {
+            assertEmptyCollection(mongodb, "test_1");
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionTest.java:46]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionTest.java (hunk 2 above)
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionTest.java:40]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBInstance.java (hunk 5 above)
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBInstance.java:42]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBTestService.java (hunk 6 above)
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBTestService.java:110]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+     */
+    public static MongoDBExtension create() {
+        return new MongoDBExtension(MongoDBTestService.DEFAULT_VERSION);
+    }
+
+    /**
+     * Create new extension instance using the given MongoDB service version.
+     *
+     * @return the new extension instance
+     */
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java:75]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionWithRegistrationAsStaticFieldTest.java (hunk 3 above)
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionWithRegistrationAsStaticFieldTest.java:39]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionWithRegistrationAsStaticFieldTest.java (hunk 4 above)
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtensionWithRegistrationAsStaticFieldTest.java:70]: :+1:

File: graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java:77]: Ah! OK, that's fine as well.

**Pull request merged**